### PR TITLE
MINIFICPP-1717 Refactor *StreamCallback to be function objects

### PR DESCRIPTION
**Pull request by szaszm**
I propose that we get rid of the *StreamCallback base classes and just use std::function to type erase the callback types where necessary. The reason is simlicity of implementation and usage.

- `InputStreamCallback` and `OutputStreamCallback` are now aliases of `std::function<int64_t(const std::shared_ptr<io::BaseStream>&)>`.
- `InputOutputStreamCallback` is now an alias of `std::function<int64_t(const std::shared_ptr<io::BaseStream>&, const std::shared_ptr<io::BaseStream>&)>`.
- `ProcessSession::read` and friends now take `const InputStreamCallback&` and `const OutputStreamCallback&`. This means value semantics of the function object, due to `std::function`. If the callback needs to preserve state after the call, it needs to be passed with `std::ref(callback)`, which wraps it in a `std::reference_wrapper<T>`.
- Introduced `ProcessSession::readBuffer` that returns the status as usual and the buffer as `std::vector<std::byte>`.
- Replaced about a dozen callback classes with calls to `readBuffer`/`writeBuffer`/`appendBuffer`.
- Replaced simple callback classes with inline lambdas or stream pipes.
- Converted stream pipes to use stream references instead of `shared_ptr`s, needed in `SiteToSiteClient`

https://issues.apache.org/jira/browse/MINIFICPP-1717

---

Thank you for submitting a contribution to Apache NiFi - MiNiFi C++.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [x] Is there a JIRA ticket associated with this PR? Is it referenced
     in the commit message?

- [x] Does your PR title start with MINIFICPP-XXXX where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [x] Has your PR been rebased against the latest commit within the target branch (typically main)?

- [x] Is your initial contribution a single, squashed commit?

### For code changes:
- [x] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?
- [x] If applicable, have you updated the LICENSE file?
- [x] If applicable, have you updated the NOTICE file?

### For documentation related changes:
- [x] Have you ensured that format looks appropriate for the output in which it is rendered?

### Note:
Please ensure that once the PR is submitted, you check GitHub Actions CI results for build issues and submit an update to your PR as soon as possible.

### CODE REVIEW

**Comment by szaszm** [comment]
I marked and unmarked this as draft because the first CI run had a C2PauseResumeTest failure on Windows. Tested multiple times on my local windows environment and couldn't reproduce it, so reran the CI and now it's green.

**Comment by lordgamez** [comment]
> I marked and unmarked this as draft because the first CI run had a C2PauseResumeTest failure on Windows. Tested multiple times on my local windows environment and couldn't reproduce it, so reran the CI and now it's green.

After checking the recent CI issues last week I added a few new flaky test issues to the following epic: [MINIFICPP-1324](https://issues.apache.org/jira/browse/MINIFICPP-1324) 
C2PauseResumeTest was one of the flaky tests which occurred in the CI environment, so it is not related to this PR, here's the relevant ticket: [MINIFICPP-1711](https://issues.apache.org/jira/browse/MINIFICPP-1711)

File: libminifi/test/unit/ContentRepositoryDependentTests.h [hunk 1]
```diff
@@ -87,20 +76,20 @@ class Fixture {
 
   core::ProcessSession &processSession() { return *process_session_; }
 
+
   void transferAndCommit(const std::shared_ptr<core::FlowFile>& flow_file) {
     process_session_->transfer(flow_file, Success);
     process_session_->commit();
   }
 
+
```
Comment (fgerlits) [inline libminifi/test/unit/ContentRepositoryDependentTests.h:85]: two new blank lines, probably added by accident

File: libminifi/include/utils/StringUtils.h [hunk 2]
```diff
@@ -402,6 +402,7 @@ class StringUtils {
     return to_hex(gsl::make_span(str).as_span<const std::byte>(), uppercase);
   }
 
+
```
[the remark this answers is not in the document]
Comment (szaszm) [reply libminifi/include/utils/StringUtils.h:405]: This one was not an accident, I thought it separated hex/base64 functions from each other better.

File: extensions/opencv/MotionDetector.cpp [hunk 3]
```diff
@@ -161,8 +160,16 @@ void MotionDetector::onTrigger(const std::shared_ptr<core::ProcessContext> &cont
   }
   cv::Mat frame;
 
-  opencv::FrameReadCallback cb(frame);
-  session->read(flow_file, &cb);
+  session->read(flow_file, [&frame](const std::shared_ptr<io::BaseStream>& inputStream) -> int64_t {
```
[the remark this answers is not in the document]
Comment (szaszm) [reply extensions/opencv/MotionDetector.cpp]: Resolved most of your comments in [46f6a43](https://github.com/apache/nifi-minifi-cpp/pull/1236/commits/46f6a43b0843948ee7c22f79c7620bf1a2a06005). Feel free to reopen the threads if you have issues with the resolution, and thanks for the review!

**Review by fgerlits — APPROVED** [review]

**Review by lordgamez — APPROVED** [review]

**Review by martinzink — APPROVED** [review]

**Pull request closed without merging**